File: c++/classes/constructors_and_destructors/constructor_with_initialisation.cpp
#include <iostream>

class Basket {
public:
  Basket();
  ~Basket();
  void put(std::string item);
  std::string pull();
  void show();
private:
  const int capacity;
  std::string *items;
  int load_counter;
};

Basket::Basket(): capacity(3) {
  items = new std::string[capacity];
  load_counter = 0;
}

Basket::~Basket() {
  delete [] items;
}

void Basket::put(std::string item) {
  if (load_counter < capacity) {
    items[load_counter] = item;
  }
  load_counter++;
}

std::string Basket::pull() {
  std::string item = items[load_counter];
  items[load_counter] = "";
  load_counter--;

  return item;
}

void Basket::show() {
  for (int i = 0; i < load_counter; i++) {
    std::cout << items[i] << " ";
  }
  std::cout << std::endl;
}

int main() {
  Basket handbasket;

  handbasket.put("apple");
  handbasket.put("pear");
  handbasket.put("plum");

  handbasket.show();
}

File: javascript/classes/constructors/constructor_with_initialisation.js
#!/usr/bin/env node

class Basket {
  constructor() {
    this.capacity = 3;
    this.load_counter = 0;
    this.items = [];
  }

  put = function(item) {
    if (this.load_counter < this.capacity) {
      this.items[this.load_counter] = item;
      this.load_counter++;
    }
  }

  pull = function() {
    let item;

    if (this.load_counter > 0) {
      item = this.items[this.load_counter - 1];
      this.items[this.load_counter - 1] = "";
      this.load_counter--;
    } else {
      item = "";
    }

    return item;
  }

  show = function() {
    for (let item of this.items) {
      process.stdout.write(item + ' ');
    }
    console.log();
  }

  capacity;
  items;
  load_counter;
}

let handbasket = new Basket();

handbasket.put("apple");
handbasket.put("pear");
handbasket.put("plum");
handbasket.put("peach");

console.log("BASKET:");

handbasket.show();

console.log("REMOVED ITEMS:");

console.log(handbasket.pull());
console.log(handbasket.pull());

console.log("BASKET:");

handbasket.show();

console.log("REMOVED ITEMS:");

console.log(handbasket.pull());
console.log(handbasket.pull());

console.log("BASKET:");

handbasket.show();
